Add support for cleanup steps

vjt opened the pull request:
This implements feature request #143.

Missing:

* handling the pause event while in cleanup steps - *DONE*
* handling errors in cleanup steps - *DONE* by ignoring them
* unit tests - *DONE*
* integration tests - *DONE*

vjt committed:
Squelch format-related warnings

Signed-off-by: Marcello Barnaba <mbarnaba@meta.com>

vjt committed:
Stub out test objects creation in a separate func

Signed-off-by: Marcello Barnaba <mbarnaba@meta.com>

vjt committed:
Cleanup steps: add data structure

The cleanup steps are embedded within a TestDescriptor. They are retrieved via
an optional separated test fetcher called "CleanupFetcherName" and with its
parameters in the "CleanupFetcherTestParameters" property.

The TestFetcherBundle structure in pkg/test/fetcher.go is changed adding the
cleanup steps alongside the existing test steps.

The pluginregistry/bundles.go change checks whether the cleanup fetcher is set
and whether its parameters are valid. If that is the case, the cleanup fetcher
and its parameters are set in the TestFetcherBundle structure.

Then, the TestStepsDescriptors structure in pkg/test/step.go is changed adding
again the cleanup name and the resolved cleanup steps alongside the existing
test steps.

The change in pkg/jobmanager/steps.go checks whether a cleanup fetcher is set
and it invokes it passing its parameter to obtain concrete cleanup steps
descriptors, that are saved in the TestStepsDescriptors structure.

The concrete cleanup steps are stored in the Test object, used at runtime,
alongside the existing test steps, and modeled as a list of TestStepBundle
structures.

The change in pkg/jobmanager/job.go checks whether cleanup steps are defined
in the TestStepsDescriptors when building the Test object, and if yes it
creates the step bundles for the cleanup steps.

The change in pkg/jobmanager/bundles.go massages the existing methods to allow
passing a list of test steps instead of a TestStepsDescriptor, so to be able to
re-use the methods for both TestSteps and CleanupSteps.

Yay for layering :-)

Signed-off-by: Marcello Barnaba <mbarnaba@meta.com>

vjt committed:
Cleanup steps: Ensure different labels vs. tests

As this is required in the next changeset where each step gets a separate event
emitter.

Signed-off-by: Marcello Barnaba <mbarnaba@meta.com>

vjt committed:
Cleanup steps: run them after test steps

This change implements the core of cleanup steps runs after test steps have completed.

It is done in a way that steps variables can be shared between test steps and cleanup
steps.

Signed-off-by: Marcello Barnaba <mbarnaba@meta.com>

mimir-d commented:
youve a broken test; 
what should i understand from the "missing" section above? option 1. you'd like this PR merged and then you add others separately (in which case, i need to check if this one is functionally complete for what it does). or option 2. this a WIP PR and you'll add more commits to it

vjt commented:
> youve a broken test;

working on fixing that

> what should i understand from the "missing" section above? option 1. you'd like this PR merged and then you add others separately (in which case, i need to check if this one is functionally complete for what it does). or option 2. this a WIP PR and you'll add more commits to it

sorry, I wrote that in a hurry. this is a WIP PR and I'll add more commits to it. I'd like your review on what it has been done so far, whether it is reasonable or not, and your insights on the missing items :-)

Thank you!

vjt committed:
Tag admin server tests separately

This allows skipping tests that rely on MongoDB, if for instance you are
running this environment on an M1 Mac with QEMU via UTM that does not yet
support the AVX machine instructions that MongoDB currently requires.

Signed-off-by: Marcello Barnaba <mbarnaba@meta.com>

vjt committed:
Remove unreachable code

Signed-off-by: Marcello Barnaba <mbarnaba@meta.com>

vjt commented on pkg/runner/job_runner.go:
```diff
@@ -454,6 +456,26 @@ func (jr *JobRunner) runTest(ctx xcontext.Context,
+			)
+			runCtx.Debugf("== test runner cleanup finished, err: %v", err)
+
+			if err == xcontext.ErrPaused {
+				return targets, testRunnerState, succeed, err
+			}
```
I don't think this will work, because we are returning the runner state of the cleanup steps runner, and when resuming this state is going to be fed to the test steps runner.

We'd need a new way to signal that we were in cleanup steps, and skip the execution of tests steps and jump directly to the cleanup steps.

mimir-d replied:
correct; you can probably make that difference by aggregating this resume state into a bigger object.
if you look at the signature `([]*target.Target, json.RawMessage, bool, error)`, it shows that the 2nd param is a json.rawmessage

so you could do, 
1. if interrupted in test steps, make another object `struct State { teststate: json.Raw, cleanupstate: json.Raw}` and return `targets, State{ teststate: testRunnerState, cleanupstate: nil }, succeed, err` (though you probably also need to move that `succeed` and `err` inside the struct
2. if interrupted in cleanup, the `State.cleanupstate` wont be nil; you may need to select on some aspect there whether test steps need to rerun at all on resume (because if we got interrupted in cleanup it means that the test has finished)

this is obviously a breaking change for the state resume format, so it'll need a json version increase (it's somewhere up where the json gets serialized). and you'll need to change it's ingestion as well

think of this in the lines of "if i get descheduled right here, how do i make the func reentrant" (kinda like how async works in python/rust)

vjt replied:
I am going through the route of adding another `json.RawMessage` attribute to the [`PauseEventPayload` struct](https://github.com/linuxboot/contest/blob/main/pkg/job/events.go#L97), calling it `CleanupRunnerState`. If that is set, it means we were in cleanup steps, so test steps are skipped. :-)

What do you think? This looks simper and should not require structure changes to the test-level `resumeStateStructVersion`.

Another point is that, in the end, this is job-level information, not test-level.

Thanks!

mimir-d replied:
ok with this design, as discussed in call

vjt commented on pkg/runner/job_runner.go:
```diff
@@ -432,15 +432,17 @@ func (jr *JobRunner) runTest(ctx xcontext.Context,
-		runCtx.Debugf("== test runner starting")
-		testRunnerState, targetsResults, err := testRunner.Run(
+		runCtx.Debugf("== test runner starting test steps")
+		testRunnerState, targetsResults, stepOutputs, err := testRunner.Run(
```
I am returning `stepOutputs` here that is the steps variable registry to feed it back when calling the cleanup steps runner, so that any variable collected during test runs is available also in the cleanup steps.

I am not sure however if that is the proper way of doing this :-)

mimir-d replied:
it may be possible to lift it from testrunner and give it as input here; if that's too much, im ok with this current form

vjt commented on tests/integ/jobmanager/common.go:
```diff
@@ -602,6 +601,44 @@ func (suite *TestJobManagerSuite) TestPauseAndFailToResume() {
+{[JOBID 1 IntegrationTest: steps 0 ][Target{ID: "id1"} TargetAcquired]}
+{[JOBID 1 IntegrationTest: steps 0 hello_world][Target{ID: "id1"} TargetIn]}
+{[JOBID 1 IntegrationTest: steps 0 hello_world][Target{ID: "id1"} TargetOut]}
+{[JOBID 1 IntegrationTest: steps 0 goodbye_world][Target{ID: "id1"} TargetIn]}
+{[JOBID 1 IntegrationTest: steps 0 goodbye_world][Target{ID: "id1"} TargetOut]}
```
these two lines should actually have `IntegrationTest: cleanup` instead of `IntegrationTest: steps` but I think this can be left as a subsequent improvement

mimir-d replied:
no no, this is an actual problem. i was saying above, if we cant discriminate between events emitted by the test steps vs events from the cleanup, it makes for a mixed mess

vjt replied:
OK, found, it was a trivial change in the TestStepEventsEmitter call, in which I was passing the test name and not the cleanup name.

vjt reviewed (commented):
Adding some comments where I think we still need work

mimir-d commented on pkg/jobmanager/bundles.go:
```diff
@@ -83,22 +84,24 @@ func newBundlesFromSteps(ctx xcontext.Context, descriptors []*test.TestStepDescr
-	}
-
-	// verify that there are not duplicated labels
+func validateNoDuplicateLables(testStepBundles []test.TestStepBundle, cleanupStepBundles []test.TestStepBundle) error {
```
typo
```suggestion
func validateNoDuplicateLabels(testStepBundles []test.TestStepBundle, cleanupStepBundles []test.TestStepBundle) error {
```

mimir-d commented on pkg/jobmanager/bundles.go:
```diff
@@ -83,22 +84,24 @@ func newBundlesFromSteps(ctx xcontext.Context, descriptors []*test.TestStepDescr
+func validateNoDuplicateLables(testStepBundles []test.TestStepBundle, cleanupStepBundles []test.TestStepBundle) error {
+	// verify that there are not duplicated labels across all test and cleanup steps.
+	// This is required later in the execution stage.
 	labels := make(map[string]bool)
```
did you know that if you use `struct{}` instead of `bool`, the map does not take any space for values? and just turns into a set, in terms of memory consumption

mimir-d commented on pkg/jobmanager/job.go:
```diff
@@ -122,6 +90,73 @@ func newJob(ctx xcontext.Context, registry *pluginregistry.PluginRegistry, jobDe
+			if err != nil {
+				return nil, fmt.Errorf("could not create cleanup test steps bundles: %w", err)
+			}
+			cleanupName = thisTestStepsDescriptors.CleanupName
```
wondering if we really need a `CleanupName`. the test name makes sense, cleanup.. not so much to me

vjt replied:
the cleanupName is used to emit events with a different prefix

mimir-d commented on pkg/jobmanager/job.go:
```diff
@@ -122,6 +90,73 @@ func newJob(ctx xcontext.Context, registry *pluginregistry.PluginRegistry, jobDe
+				return nil, err
+			}
+		}
+		if err := validateNoDuplicateLables(bundleTest, bundleCleanup); err != nil {
```
you can remove duplication here (and an extra loop in func) by saying
```suggestion
		if err := validateNoDuplicateLables(append(bundleTest, bundleCleanup...)); err != nil {
```

mimir-d commented on pkg/pluginregistry/bundles.go:
```diff
@@ -60,9 +61,29 @@ func (r *PluginRegistry) NewTestFetcherBundle(ctx xcontext.Context, testDescript
+	// Initialization and validation of the optional CleanupFetcher and its parameters
+	var (
+		cleanupFetcher test.TestFetcher
+		cp             interface{}
```
`cp`? 

it's better if we dont follow this old C style, generally speaking. ok in short loops or short funcs. this is neither

vjt replied:
OK, fixed also the `fp` -> `fetchParameters`

mimir-d commented on pkg/runner/job_runner.go:
```diff
@@ -454,6 +456,26 @@ func (jr *JobRunner) runTest(ctx xcontext.Context,
+				runCtx,
+				t,
+				targets,
+				NewTestStepEventsEmitterFactory(jr.storageEngineVault, j.ID, runID, t.Name, testAttempt),
```
this is probably not a good idea; the events emitted in cleanup should be separated from test events;
see if there's some aspect to differentiate between them (you can run a test job where both test and cleanup steps emit events; cmd/exec plugin should be simplest to get the stdout event and check in the output status json)

vjt replied:
we can distinguish them easily by using the Cleanupname instead of the test Name. What do you think?

mimir-d commented on pkg/runner/test_runner.go:
```diff
@@ -137,11 +139,11 @@ func (tr *TestRunner) Run(
 	}
 
-	stepOutputs, err := newTestStepsVariables(t.TestStepsBundles, func(tgtID string, stepLabel string, name string, value json.RawMessage) {
+	onVariableAdded := func(tgtID string, stepLabel string, name string, value json.RawMessage) {
```
```suggestion
	onStepOutputAdded := func(tgtID string, stepLabel string, name string, value json.RawMessage) {
```

this was supposed to be named "step outputs" in the first place

mimir-d commented on pkg/runner/test_runner.go:
```diff
@@ -160,22 +162,27 @@ func (tr *TestRunner) Run(
+			ctx.Errorf("Failed to initialise test steps variables: %v", err)
+			return nil, nil, nil, err
+		}
+		stepOutputs = stepOutputs_
```
ehh?

just dont define a new var? `stepOutputs, err = newTestStepsVariables...`

mimir-d commented on pkg/runner/test_runner.go:
```diff
@@ -160,22 +162,27 @@ func (tr *TestRunner) Run(
-		if err := stepOutputs.initTargetStepsVariables(targetID, targetState.StepsVariables); err != nil {
-			ctx.Errorf("Failed to initialise test steps variables for target: %s: %v", targetID, err)
-			return nil, nil, err
+	if stepOutputs == nil {
```
kinda not pretty; as i was saying above, see if you can lift this

vjt replied:
I think I can do that, but in a subsequent changeset, to not further complicate this one

mimir-d commented on pkg/test/step.go:
```diff
@@ -72,8 +72,10 @@ type TestStepLoader func() (string, TestStepFactory, []event.Name)
-	TestSteps []*TestStepDescriptor
+	TestName     string
+	TestSteps    []*TestStepDescriptor
+	CleanupName  string
```
mentioned prev; i dont think we need a name for the cleanup; it's part of the test and the test has a name already

vjt replied:
as mentioned, we want a way to distinguish the events emitted by the test with the ones emitted by the cleanup :-)

mimir-d commented:
please also change this to merge into `develop` instead of `main`

vjt committed:
Tests: allow retrieving events for more than one test

Signed-off-by: Marcello Barnaba <mbarnaba@meta.com>

vjt committed:
Cleanup steps: do not require a cleanup test name

As the cleanup steps are strictly related to the test they refer to, remove the
requirement to be in a named test.

Signed-off-by: Marcello Barnaba <mbarnaba@meta.com>

mimir-d commented on plugins/storage/rdbms/events.go:
```diff
@@ -100,9 +100,9 @@ func buildTestEventQuery(baseQuery safesql.TrustedSQLString, testEventQuery *tes
 	}
 
-	if testEventQuery.TestName != "" {
+	if testEventQuery.TestNames != nil {
```
nit; go doesnt make much of a difference between an empty slice and a nil slice; you can iterate on both
however, your check here only looks for nil. would be better as `if len(...) > 0`

mimir-d commented on pkg/runner/job_runner_test.go:
```diff
@@ -283,8 +281,8 @@ func (s *JobRunnerSuite) TestJobWithTestRetry() {
 {[1 1 SimpleTest 0 test_step_label][Target{ID: "T1"} TargetErr &"{\"Error\":\"some error\"}"]}
-{[1 1 SimpleTest:CLEANUP 0 test_cleanup_step][Target{ID: "T1"} TargetIn]}
-{[1 1 SimpleTest:CLEANUP 0 test_cleanup_step][Target{ID: "T1"} TargetOut]}
+{[1 1 SimpleTest: CLEANUP 0 test_cleanup_step][Target{ID: "T1"} TargetIn]}
```
should there be extra spaces here? thought the discussion result was "%s:CLEANUP"

vjt replied:
I've decided to add the extra space because of [this test](https://github.com/vjt/contest/blob/feature/143/tests/integ/jobmanager/common.go#L635-L636) as `:` is not a special char and users can use it. What do you think?

mimir-d replied:
right; no good character, at the end of the day it doesnt really matter what that is as long as it's not too confusing. feel free to keep as is, or without the space, or another separator like `@%#`

vjt replied:
I might end up with ` [CLEANUP]`

mimir-d reviewed (changes requested):
not many changes, but one of the tests is failing

vjt commented:
Tests finally pass! [0f7af7c](https://github.com/linuxboot/contest/pull/163/commits/0f7af7cce78f8c973fb461f0b7866fa1ef1f2639) was particularly painful, because I was thinking to make a change affecting only test code, but in reality the code that fetches from storages is the same across tests and prod. Anyway, the change looks reasonable to me.

Cheers

mimir-d commented:
thanks for working on this
